docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+It's probably worth noting here that the [Result](https://doc.rust-lang.org/std/result/) type itself is just a normal enum in Rust with two values (`Ok` and `Err`).
+
+## Matching
+
+### Prefer match statements for exhaustive conditional code :: suggested
+
+Matching is Rust's idiomatic way to handle any non-trivial exhaustive conditional code, and works elegantly with enums, but handles any data for which a predicate can be formed (which is basically everything).
+
+Even a simple if/else block can be expressed more idiomatically using a match statement and will produce essentially the same compiled code after optimization. This is especially true if you want all cases to be covered, since this is enforced by a match statement.
+
+From [Rust By Example](https://doc.rust-lang.org/rust-by-example/flow_control/match.html):
+
+```
+match number {
+    // Match a single value
+    1 => println!("One!"),
+    // Match several values
+    2 | 3 | 5 | 7 | 11 => println!("This is a prime"),
+    // Match an inclusive range
+    13..=19 => println!("A teen"),
+    // Handle the rest of cases
+    _ => println!("Ain't special"),
+}
+```
+
+Another nice property of match is that it returns a value, which tied to the fact that it must cover all cases means you get a readable idiom for ad-hoc matching and mapping:
+
+```
+let x = match number {
+    0     => "Zero",
+    1..=9 => "Some",
+    _     => "Many",
+}
+```
+
+Obviously where an if-statement is simply there to do optional work, and not cover every case, it may well be more suitable to just use that.
+
+# Error Handling
+
+See also the [Error Handling](https://doc.rust-lang.org/book/ch09-00-error-handling.html) chapter in the Rust Book.
+
+The ICU4X library should be designed so as to never fail at runtime. Now obviously that's something you'd expect all library writers to say, but in Rust you can control the places where code can fail explicitly, so it's much easier to write self-documenting code where the assumptions around failure are obvious.
+
+Most core rust APIs (traits) have two ways to access data, a version that can "panic" (equivalent to a SEGFAULT in C/C++), and one which returns a [Result](https://doc.rust-lang.org/std/result/index.html) or an [Option](https://doc.rust-lang.org/std/option/enum.Option.html). Rust is a language which likes to avoid "unnecessary" overhead, and in some cases it is perfectly correct to use an API which can "panic" because you have already checked the arguments carefully (and performance will be better).
+
+Note that in cases where the Rust compiler can statically determine that a check is sufficient to avoid panic, it will remove the internal check and panic related code, leaving just a provably safe data access.
+
+**Open Question**: Should we have a standard `IcuResult<T>` type (e.g. `Result<T, &'static str>`, or something which could later provide internationalized error messages; e.g. an enum per error).
```

Comment:
probably yeah

we can use an extensible enum so that we can keep adding variants to it

Reply:
@steveklabnik in https://news.ycombinator.com/item?id=22957878 :
```
The current error handling consensus seems to be "anyhow for applications, either no crate or thiserror for libraries." Obviously this is not uniform, but the discussion + download counts point to this, imho. 
```